### PR TITLE
Alternative Implementation of "Use Stiffness on Old Iteration"

### PR DESCRIPTION
**Pull request by RiccardoRossi**

### CODE REVIEW

File: kratos/solving_strategies/builder_and_solvers/builder_and_solver.h [hunk 1]
```diff
@@ -429,6 +429,27 @@ class BuilderAndSolver
...
+    /**
+     * @brief Function to perform the building and solving phase at the same time Linearizing with the database at the old iteration
+     * @details It is ideally the fastest and safer function to use when it is possible to solve just after building
+     * @param pScheme The pointer to the integration scheme
+     * @param rModelPart The model part to compute
+     * @param rA The LHS matrix of the system of equations
+     * @param rDx The vector of unkowns
+     * @param rb The RHS vector of the system of equations
+     */
+    virtual void BuildAndSolve_LinearizedOnOldIteration(
```
Comment (loumalouomega) [inline kratos/solving_strategies/builder_and_solvers/builder_and_solver.h]: I don't like the name :P -> BuildAndSolveLinearizedOnPreviousIteration

File: kratos/solving_strategies/builder_and_solvers/builder_and_solver.h [hunk 2]
```diff
@@ -429,6 +429,27 @@ class BuilderAndSolver
...
+        typename TSchemeType::Pointer pScheme,
+        ModelPart &rModelPart,
+        TSystemMatrixType &rA,
+        TSystemVectorType &rDx,
+        TSystemVectorType &rb,
+        bool mesh_moving_needed)
```
Comment (loumalouomega) [inline kratos/solving_strategies/builder_and_solvers/builder_and_solver.h]:
```suggestion
        const bool MeshMovingNeeded)
```

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h [hunk 3]
```diff
@@ -178,7 +175,7 @@ class ResidualBasedBlockBuilderAndSolver
         // Getting the array of the conditions
         const int nconditions = static_cast<int>(rModelPart.Conditions().size());
 
-        ProcessInfo& CurrentProcessInfo = rModelPart.GetProcessInfo();
+        ProcessInfo &CurrentProcessInfo = rModelPart.GetProcessInfo();
```
Comment (loumalouomega) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]: Why moving the &, BTW this is an aesthetic change

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h [hunk 4]
```diff
@@ -263,8 +259,8 @@ class ResidualBasedBlockBuilderAndSolver
      */
     void BuildLHS(
         typename TSchemeType::Pointer pScheme,
-        ModelPart& rModelPart,
-        TSystemMatrixType& A) override
+        ModelPart &rModelPart,
```
Comment (loumalouomega) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]: Moving the & is unnecessary and it is an aesthetic change (usually if I do that I get blocked). When we decided to move the &? BTW

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h [hunk 5]
```diff
@@ -378,14 +374,17 @@ class ResidualBasedBlockBuilderAndSolver
         else
             norm_b = 0.00;
 
-        if (norm_b != 0.00) {
+        if (norm_b != 0.00)
```
Comment (loumalouomega) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]: This is a comestic change and does not follows the styling guide { in the same line

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h [hunk 6]
```diff
@@ -378,14 +374,17 @@ class ResidualBasedBlockBuilderAndSolver
...
             //provide physical data as needed
-            if(BaseType::mpLinearSystemSolver->AdditionalPhysicalDataIsNeeded() )
+            if (BaseType::mpLinearSystemSolver->AdditionalPhysicalDataIsNeeded())
                 BaseType::mpLinearSystemSolver->ProvideAdditionalData(A, Dx, b, BaseType::mDofSet, rModelPart);
 
             //do solve
             BaseType::mpLinearSystemSolver->Solve(A, Dx, b);
-        } else {
+        }
+        else
```
Comment (loumalouomega) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]: Idem, why these changes?

**Comment by loumalouomega** [comment]
BTW, this branch totally conflicts with master

**Comment by loumalouomega** [comment]
> BTW, this branch totally conflicts with master

Fixed

File: kratos/solving_strategies/builder_and_solvers/builder_and_solver.h [hunk 7]
```diff
@@ -429,6 +429,29 @@ class BuilderAndSolver
...
+     * @param rb The RHS vector of the system of equations
+     * @param MeshMovingNeeded tells if the update of the scheme needs  to be performed when calling the Update of the scheme
+     */
+    virtual void BuildAndSolve_LinearizedOnOldIteration(
+        typename TSchemeType::Pointer pScheme,
+        ModelPart& rModelPart,
+        TSystemMatrixType& rA,
+        TSystemVectorType& rDx,
+        TSystemVectorType& rb,
+        const bool MeshMovingNeeded
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/builder_and_solver.h]: Why not MoveMesh like we usually name it?

File: kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h [hunk 8]
```diff
@@ -1186,6 +1214,23 @@ class ResidualBasedNewtonRaphsonStrategy
         mMaxIterationNumber = Settings["max_iterations"].GetInt();
         mReformDofSetAtEachStep = Settings["reform_dofs_at_each_step"].GetBool();
         mCalculateReactionsFlag = Settings["calculate_reactions"].GetBool();
+        mUseOldStiffnessInFirstIteration = Settings["use_old_stiffness_in_first_iteration"].GetBool();
+
+        if(mUseOldStiffnessInFirstIteration)
+        {
+            if( BaseType::GetModelPart().GetBufferSize() == 1)
+            {
+                KRATOS_WARNING("NR-Strategy")
```
Comment (philbucher) [inline kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h]: Why warning and not error?
We usually throw if the input is incorrect

File: kratos/solving_strategies/builder_and_solvers/builder_and_solver.h (hunk 1 above)
Comment (RiccardoRossi) [reply kratos/solving_strategies/builder_and_solvers/builder_and_solver.h]: ok

File: kratos/solving_strategies/builder_and_solvers/builder_and_solver.h (hunk 7 above)
Comment (RiccardoRossi) [reply kratos/solving_strategies/builder_and_solvers/builder_and_solver.h]: ok. Actually my intention was not to to change the interface and pass a variable through the processinfo, however that i realized that the processinfo may be shared between multiple modelparts and hence multiple strategies, so essentially it is needed to pass it this way

File: kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h (hunk 8 above)
Comment (RiccardoRossi) [reply kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h]: you are right, i removed this from here and added an error from within the builder and solver.

File: kratos/python/add_strategies_to_python.cpp [hunk 9]
```diff
@@ -461,6 +462,7 @@ namespace Kratos
             py::class_< ResidualBasedBlockBuilderAndSolverType, ResidualBasedBlockBuilderAndSolverType::Pointer,BuilderAndSolverType>(m,"ResidualBasedBlockBuilderAndSolver")
             .def(py::init< LinearSolverType::Pointer > ())
             .def(py::init< LinearSolverType::Pointer, Parameters > ())
+            .def("BuildAndSolve_LinearizedOnOldIteration", &ResidualBasedBlockBuilderAndSolverType::BuildAndSolve_LinearizedOnOldIteration)
```
Comment (RiccardoRossi) [inline kratos/python/add_strategies_to_python.cpp]: this is to be removed (note for myself)

**Comment by RiccardoRossi** [comment]
@loumalouomega you are right, i did not intend to do any cosmetic change. (and definitely not to move the positon of the &)

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py [hunk 10]
```diff
@@ -495,14 +496,17 @@ def _create_newton_raphson_strategy(self):
                                                                      self.settings["compute_reactions"].GetBool(),
                                                                      self.settings["reform_dofs_at_each_step"].GetBool(),
                                                                      self.settings["move_mesh_flag"].GetBool())
+        #strategy.SetUseOldStiffnessInFirstIterationFlag(self.settings["use_old_stiffness_in_first_iteration"].GetBool())
+        strategy.SetUseOldStiffnessInFirstIterationFlag(True)
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py]: comment for myself. the option on the line just before is the one that should ultimately go in.

i am forcing it to true to ensure that all the tests pass.

File: applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py [hunk 11]
```diff
@@ -376,23 +378,40 @@ def test_compare_TL_UL_2D_triangle(self):
...
             for i in range(ul_lhs.Size1()):
                 for j in range(ul_lhs.Size2()):
-                    self.assertLess((ul_lhs[i, j] - tl_lhs[i, j]) / tl_lhs[i, j], 1.0e-3)
+                    self.assertLess(abs(ul_lhs[i, j] - tl_lhs[i, j]) / abs(tl_lhs[i, j]), 1.0e-10)
 
         #self.__post_process(tl_mp)
         #self.__post_process(ul_mp)
 
-    def test_TL_2D_triangle(self):
+    def test_compare_TL_UL_2D_triangle_block(self):
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:393]: @loumalouomega @philbucher i changed this starting from #6572 but pusing it further.

one bad thing is that the UL hexa test fails for the level of deformation which is originally used in the test.
The TL test does not fail for the same level of deformation, which implies that we have some problem with the position used as reference in the UL, something that i believe is very related to #6567

File: applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py [hunk 12]
```diff
@@ -460,17 +497,35 @@ def test_TL_2D_quadrilateral(self):
...
 
-    def test_TL_3D_tetra(self):
+    def test_TL_2D_quadrilateral_block(self):
+        builder_type = "block_builder"
+        linearize_on_old_iteration = False
+        self._TL_2D_quadrilateral(builder_type, linearize_on_old_iteration,6 )
+
+    def test_TL_2D_quadrilateral_linearized_on_old_iteration(self):
+        builder_type = "block_builder"
+        linearize_on_old_iteration = True
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:514]: Shouldn’t it require less iterations in this case?

**Review by RiccardoRossi — COMMENTED** [review]
@marandra @SergioJimenezReyes @AlejandroCornejo can u please try if your tests run on this branch?

File: applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py (hunk 12 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:514]: not necessarily....  the linearization on the previous iteration may provide a better or worst starting point. It will be typically better if it is an increment wrt the previous iteration (as happens in real cases) but not if all deformation is applied at once (as it is the case here)

File: applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py (hunk 11 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:393]: @RiccardoRossi the element is inverted, so it fails, it is not an error of the UL element:

imagen

imagen

The TL integrates in the undeformed configuration, so it does not crash.

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:393]: BTW I obtained this hardcoding in the NR: 

imagen

Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:393]: Honestly i did not doubt that the element got inverted.

the problem is the WHY it gets inverted. One of the checks you do is  that the LHS and RHS are identical (as they should be) for TL and UL. The problem is that the TL test passes and the UL does not ... meaning that at some point the two matrices must be different. no?

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:393]: TL integrates in the undeformed position, therefore these jacobian is constant. If after solving the system the configuration is not broken the problem is solved. UL integrates in the deformed configuration, this means that is not possible to integrate the volume if that configuration is inverted. Maybe removing the error line the result is the same, but this is not the point. You may reach the same solution going few by few without breaking the element.

File: applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py (hunk 12 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:514]: well, after the last fix it does. (Apparently always 2 with the new implementation)

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py [hunk 13]
```diff
@@ -511,3 +515,6 @@ def _create_line_search_strategy(self):
                                                      self.settings["compute_reactions"].GetBool(),
                                                      self.settings["reform_dofs_at_each_step"].GetBool(),
                                                      self.settings["move_mesh_flag"].GetBool())
+        #strategy.SetUseOldStiffnessInFirstIterationFlag(self.settings["use_old_stiffness_in_first_iteration"].GetBool())
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py]: comment for myself: same here, the good line is the commente one

File: applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py (hunk 11 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/tests/test_patch_test_large_strain.py:393]: there was an error in my implementation. now it is solved

**Review by RiccardoRossi — COMMENTED** [review]
ok, i think this is now ready for merge
@KratosMultiphysics/technical-committee 

on a side note, i would argue that the current default parameters of line search are not tuned for robustness, but i will open a separate issue about that

**Comment by philbucher** [comment]
@armingeiser I think this can be interesting for your meshmoving evaluation project

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py [hunk 14]
```diff
@@ -495,14 +496,16 @@ def _create_newton_raphson_strategy(self):
...
+        return strategy
 
     def _create_line_search_strategy(self):
         computing_model_part = self.GetComputingModelPart()
         mechanical_scheme = self.get_solution_scheme()
         linear_solver = self.get_linear_solver()
         mechanical_convergence_criterion = self.get_convergence_criterion()
         builder_and_solver = self.get_builder_and_solver()
-        return KratosMultiphysics.LineSearchStrategy(computing_model_part,
+        strategy = KratosMultiphysics.LineSearchStrategy(computing_model_part,
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:508]: why this change?

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h [hunk 15]
```diff
@@ -488,6 +488,110 @@ class ResidualBasedBlockBuilderAndSolver
...
+
+        // Use UpdateDatabase to bring back the solution to how it was at the end of the previous step
+        pScheme->Update(rModelPart, BaseType::mDofSet, rA, dx_prediction, rb);
+        const auto it_node_begin = rModelPart.NodesBegin();
+        if (MoveMesh) {
+            #pragma omp parallel for
+            for (int i = 0; i < static_cast<int>(rModelPart.NumberOfNodes()); ++i) {
+                auto it_node = it_node_begin + i;
+                noalias(it_node->Coordinates()) = it_node->GetInitialPosition().Coordinates()
+                                                + it_node->FastGetSolutionStepValue(DISPLACEMENT);
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]: hmmmmm
Hardcoding variables in the B&S ...?

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py [hunk 16]
```diff
@@ -112,6 +112,7 @@ def GetDefaultSettings(cls):
             "displacement_control": false,
             "reform_dofs_at_each_step": false,
             "line_search": false,
+            "use_old_stiffness_in_first_iteration": false,
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:115]: what happens if this is used in combination with a linear analysis?

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h [hunk 17]
```diff
@@ -679,9 +783,6 @@ class ResidualBasedBlockBuilderAndSolver
 
         KRATOS_INFO_IF("ResidualBasedBlockBuilderAndSolver", ( this->GetEchoLevel() > 2 && rModelPart.GetCommunicator().MyPID() == 0)) << "Finished setting up the dofs" << std::endl;
 
-        KRATOS_INFO_IF("ResidualBasedBlockBuilderAndSolver", ( this->GetEchoLevel() > 2)) << "End of setup dof set\n" << std::endl;
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h:682]: can you please make a separate PR for changes unrelated to this one?

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h [hunk 18]
```diff
@@ -488,6 +488,110 @@ class ResidualBasedBlockBuilderAndSolver
...
+            KRATOS_ERROR << "BlockBuilderAndSolver: \n"
+                << "The buffer size needs to be at least 2 in order to use \n"
+                << "BuildAndSolve_LinearizedOnOldIteration \n"
+                << "current buffer size for modelpart: " << rModelPart.Name() << std::endl
+                << "is :" << rModelPart.GetBufferSize()
+                << " Old Stiffness on First Iteration. \n"
+                << "setting mUseOldStiffnessInFirstIteration=false " << std::endl;
+        }
+
+        DofsArrayType FixedDofs;
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]:
```suggestion
        DofsArrayType fixed_dofs;

```

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h [hunk 19]
```diff
@@ -488,6 +488,110 @@ class ResidualBasedBlockBuilderAndSolver
...
+        // Use UpdateDatabase to bring back the solution
+        // to where it was taking into account BCs
+        // it is done here so that constraints are correctly taken into account right after
+        pScheme->Update(rModelPart, BaseType::mDofSet, rA, dx_prediction, rb);
+        if (MoveMesh) {
+            #pragma omp parallel for
+            for (int i = 0; i < static_cast<int>(rModelPart.NumberOfNodes()); ++i) {
+                auto it_node = it_node_begin + i;
+                noalias(it_node->Coordinates()) = it_node->GetInitialPosition().Coordinates()
+                                                + it_node->FastGetSolutionStepValue(DISPLACEMENT);
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]: Aside from the hardcoding, I think there is a fct for this in the variableutils

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h [hunk 20]
```diff
@@ -488,6 +488,110 @@ class ResidualBasedBlockBuilderAndSolver
...
+        TSparseSpace::UnaliasedAdd(rb, -1.0, rhs_addition);
+
+        for(auto& dof : FixedDofs)
+            dof.FixDof();
+
+        if (!rModelPart.MasterSlaveConstraints().empty()) {
+            this->ApplyConstraints(pScheme, rModelPart, rA, rb);
+        }
+        this->ApplyDirichletConditions(pScheme, rModelPart, rA, rDx, rb);
+        // TODO: Here we should use SystemSolveWithPhysics
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]: seems no longer relevant

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h [hunk 21]
```diff
@@ -488,6 +488,110 @@ class ResidualBasedBlockBuilderAndSolver
...
+        TSystemMatrixType& rA,
+        TSystemVectorType& rDx,
+        TSystemVectorType& rb,
+        const bool MoveMesh
+        ) override
+    {
+        KRATOS_INFO_IF("BlockBuilderAndSolver", this->GetEchoLevel() > 0)
+            << "Linearizing on Old iteration" << std::endl;
+
+        if(rModelPart.GetBufferSize() == 1) {
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]: KRATOS_ERROR_IF

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h [hunk 22]
```diff
@@ -488,6 +488,110 @@ class ResidualBasedBlockBuilderAndSolver
...
+            KRATOS_ERROR << "BlockBuilderAndSolver: \n"
+                << "The buffer size needs to be at least 2 in order to use \n"
+                << "BuildAndSolve_LinearizedOnOldIteration \n"
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]:
```suggestion
                << "BuildAndSolveLinearizedOnPreviousIteration \n"

```

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h [hunk 23]
```diff
@@ -488,6 +488,110 @@ class ResidualBasedBlockBuilderAndSolver
...
+                << "current buffer size for modelpart: " << rModelPart.Name() << std::endl
+                << "is :" << rModelPart.GetBufferSize()
+                << " Old Stiffness on First Iteration. \n"
+                << "setting mUseOldStiffnessInFirstIteration=false " << std::endl;
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]: this is not an option of this class

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py (hunk 16 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:115]: it works without any problem. in that case it would be exact (note however that we would need to change the linear_strategy which actually would not be a bad idea)

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py (hunk 14 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:508]: because i need to call the constructor and then the set function.

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h (hunk 15 above)
Comment (RiccardoRossi) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]: DISPLACEMENT is a special case (and it has always being treated as such)

we always assume that X = X0+DISPLACEMENT_X (the move mesh of the strategy always did that)

**Comment by marandra** [comment]
@RiccardoRossi  Just to mention that my tests are passing now. Thanks.

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h (hunk 21 above)
Comment (RiccardoRossi) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]: ok

**Comment by SergioJimenezReyes** [comment]
> @RiccardoRossi Just to mention that my tests are passing now. Thanks.

Mine pass too.

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h (hunk 23 above)
Comment (RiccardoRossi) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]: I think this check belongs here, but i changed the error message to tell that the settings of the strategy needs to be adjusted

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h (hunk 22 above)
Comment (RiccardoRossi) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]: done

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h (hunk 19 above)
Comment (RiccardoRossi) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]: Something that could be done, but is a much larger change, is to have the function MoveMesh in the base scheme instead of in the strategy. That way one may reimplement it.

However as i am telling this is a large change (it would require revising all the schemes) so i really would not like doing it as a part of this PR

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h (hunk 17 above)
Comment (RiccardoRossi) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h:682]: I could ... but this is doing exactly the same as in line 680 just above except that line 680 already has a correction for MPI which is not present here. So... i would make an exception for this specifically.

Comment (RiccardoRossi) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h:682]: Also i think this is a merge feature: @roigcarlo corrected the line above, and the merge did not recognize that the two lines should one replace the other

**Review by RiccardoRossi — COMMENTED** [review]
ok, to me this is ready to go in.

get ready as after R8.0 i will suggest activating this behaviour by default (for now it defaults to the old behaviour)

**Comment by philbucher** [comment]
(I haven’t tried this yet)
I don’t have objections making this the default if it is as good as you say
However since this is only implemented for the Block-B&S this should only be used by default if one uses this B&S 
Otherwise we will get LOTS of warnings suddenly for the other B&Ss (i.e. the elimination and all trilinos-B&S)

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py (hunk 14 above)
Comment (philbucher) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:508]: but why not passing it like the other params?

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h (hunk 17 above)
Comment (philbucher) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h:682]: ok

**Review by philbucher — COMMENTED** [review]
minor comments abt the implementation, otherwise looks good to me now
TBD is the issue with the warning, but this can also be discussed in a future PR

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py (hunk 16 above)
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:115]: Please use the structure for the parameters we agreed in #5939

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py [hunk 24]
```diff
@@ -486,7 +487,7 @@ def _create_newton_raphson_strategy(self):
         linear_solver = self.get_linear_solver()
         mechanical_convergence_criterion = self.get_convergence_criterion()
         builder_and_solver = self.get_builder_and_solver()
-        return KratosMultiphysics.ResidualBasedNewtonRaphsonStrategy(computing_model_part,
+        strategy = KratosMultiphysics.ResidualBasedNewtonRaphsonStrategy(computing_model_part,
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:490]: Please use Parameters as input instead of adding more and more arguments to the constructor

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py (hunk 14 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:508]: Idem as my previous comment, and see @philbucher  comment

File: kratos/solving_strategies/builder_and_solvers/builder_and_solver.h [hunk 25]
```diff
@@ -429,6 +429,29 @@ class BuilderAndSolver
...
+    virtual void BuildAndSolveLinearizedOnPreviousIteration(
+        typename TSchemeType::Pointer pScheme,
+        ModelPart& rModelPart,
+        TSystemMatrixType& rA,
+        TSystemVectorType& rDx,
+        TSystemVectorType& rb,
+        const bool MoveMesh
+        )
+    {
+        KRATOS_WARNING("BuilderAndSolver") << "no special implementation available for BuildAndSolveLinearizedOnPreviousIteration. Using normal BuildAndSolve" << std::endl;
```
Comment (loumalouomega) [inline kratos/solving_strategies/builder_and_solvers/builder_and_solver.h]:
```suggestion
        KRATOS_WARNING("BuilderAndSolver") << "No special implementation available for BuildAndSolveLinearizedOnPreviousIteration. Using normal BuildAndSolve" << std::endl;
```

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h [hunk 26]
```diff
@@ -488,6 +489,96 @@ class ResidualBasedBlockBuilderAndSolver
...
+        KRATOS_ERROR_IF(rModelPart.GetBufferSize() == 1) << "BlockBuilderAndSolver: \n"
+                << "The buffer size needs to be at least 2 in order to use \n"
+                << "BuildAndSolveLinearizedOnPreviousIteration \n"
+                << "current buffer size for modelpart: " << rModelPart.Name() << std::endl
+                << "is :" << rModelPart.GetBufferSize()
+                << " Please set IN THE STRATEGY SETTINGS "
+                << " UseOldStiffnessInFirstIteration=false " << std::endl;
+
+        DofsArrayType fixed_dofs;
+        for(auto& dof : BaseType::mDofSet){
```
Comment (loumalouomega) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]:
```suggestion
        for(auto& r_dof : BaseType::mDofSet){

```
Idem for the rest

File: kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h [hunk 27]
```diff
@@ -1084,6 +1105,12 @@ class ResidualBasedNewtonRaphsonStrategy
      */
     bool mCalculateReactionsFlag;
 
+    /**
+     * @brief Flag telling if a full update of the database will be performed at the first iteration
+     * @details default = false
+     */
+    bool mUseOldStiffnessInFirstIteration = false;
```
Comment (loumalouomega) [inline kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h:1112]: I would use a flag instead of a bool

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Some comments

**Review by loumalouomega — APPROVED** [review]

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py (hunk 16 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:115]: My only remaining concerns are these

**Review by loumalouomega — CHANGES_REQUESTED** [review]

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py (hunk 14 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:508]: If i change the constructor, i will break all current usages in which this is not passed.

As i understand the idea is to pass soon to the factory and hence to passing the parameters directly, so i wouldn't propagate this change through the code to then change it agan as soon as we pass to the parameters construction. 
This is particularly true since we are very short away from the Release, so i really would avoid pervasive changes.

Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:508]: (note that using it as default is not enough, since all the previous parameters are given which forces to prescribe the last one)

**Review by loumalouomega — DISMISSED** [review]

**Comment by loumalouomega** [comment]
OK, unblocking

File: kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h (hunk 27 above)
Comment (RiccardoRossi) [reply kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h:1112]: it is a question of taste...i don't see why not using a bool here (anyway we are not passing a flag from outside)

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h (hunk 26 above)
Comment (RiccardoRossi) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]: ok

File: kratos/solving_strategies/builder_and_solvers/builder_and_solver.h [hunk 28]
```diff
@@ -448,7 +448,7 @@ class BuilderAndSolver
         const bool MoveMesh
         )
     {
-        KRATOS_WARNING("BuilderAndSolver") << "no special implementation available for BuildAndSolveLinearizedOnPreviousIteration. Using normal BuildAndSolve" << std::endl;
+        KRATOS_WARNING("BuilderAndSolver") << "No special implementation available for BuildAndSolveLinearizedOnPreviousIteration. Using normal BuildAndSolve" << std::endl;
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/builder_and_solver.h]: @loumalouomega  thx for making the changes!!

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py (hunk 24 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:490]: I agree about using parameters, however for now it is not used.
As soon as the factories are in i will also push for removing the other constructors

**Comment by RiccardoRossi** [comment]
@roigcarlo as a follow up we need to do the same for the trilinos_block_builder_and_solver (i would do in a separate PR). I'll need your help for that, but i think it will be relatively straightforward

File: kratos/solving_strategies/builder_and_solvers/builder_and_solver.h (hunk 28 above)
Comment (philbucher) [reply kratos/solving_strategies/builder_and_solvers/builder_and_solver.h]: Are you sure this warning is even necessary?
I mean, there is no harm in not having this implementation in a derived class except maybe slower convergence

Comment (RiccardoRossi) [reply kratos/solving_strategies/builder_and_solvers/builder_and_solver.h]: I think it is important to know that an option you prescribe is ignored...

actually i think there is a case for throwin an error...

Comment (philbucher) [reply kratos/solving_strategies/builder_and_solvers/builder_and_solver.h]: Throwing seems like appropriate in this situation

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h [hunk 29]
```diff
@@ -488,6 +489,96 @@ class ResidualBasedBlockBuilderAndSolver
...
+        for (int i = 0; i < static_cast<int>(BaseType::mDofSet.size()); ++i) {
+            auto it_dof = it_dof_begin + i;
+            //NOTE: this is initialzed to - the value of dx prediction
+            dx_prediction[it_dof->EquationId()] = -(it_dof->GetSolutionStepValue() - it_dof->GetSolutionStepValue(1));
+        }
+
+        // Use UpdateDatabase to bring back the solution to how it was at the end of the previous step
+        pScheme->Update(rModelPart, BaseType::mDofSet, rA, dx_prediction, rb);
+        if (MoveMesh) {
+            VariableUtils().UpdateCurrentPosition(rModelPart.Nodes(),DISPLACEMENT,0);
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h:550]: Just for the record, I think if we want to use this with the Mesh-Motion solvers we need to make this flexible (sort of like the `MoveMesh` function of the Strategy) bcs those use `MESH_DISPLACEMENT`

**Review by philbucher — APPROVED** [review]
@KratosMultiphysics/technical-committee approves this 👍 

please remember to clean/close the issues and PRs related to this
